docs/proposals/kubectl-apply-purge-from-namespace.md

```diff
@@ -0,0 +1,103 @@
+# Enable purging resources in `kubectl apply`
+
+## Motivation
+
+See [#19805](https://github.com/kubernetes/kubernetes/issues/19805).
+
+Currently `kubectl apply` will create any new resources that exist in the resource file(s) it is supplied with, but it has no way of deleting anything that has been removed from such files.
+
+This makes it difficult to use `kubectl apply` to update reality to match the declarative definitions of resources when that definition evolves over time, since changes can include deletions.
+
+The primary driver for this use-case is to be able to manage cluster add-ons using `kubectl apply`.
+
+## Proposed Solution
+
+In a "simplest possible idea first" approach, this proposal suggests that we reuse namespaces as a de-facto way of grouping resources for the purpose of _purging_ (removing) resources which no longer exist.
+
+Most add-ons reside in `kube-system` namespace, therefore for add-ons, an implementation based on namespaces would be reasonable.
+The benefit of this is that no new abstraction (such as a new way of grouping resources) is required for this use-case.
+
+Therefore, the proposal is to introduce an optional `--purge-from-namespace` flag to `kubectl apply`, which will allow a user to automatically purge resources from the given namespace, if they haven't been explicitly declared in files that are arguments to the same `kubectl apply` invocation.
+
+All resources referred to must be explicitly labelled with one non-default namespace in order for the `--purge-from-namespace` argument to succeed.
+
+This flag will require `--namespace`, and will not accept namespaces set in the input resource definitions.
```

Comment:
I think it would be reasonable to allow the namespace from the context to be used. 

I agree that allowing other namespaces in the resource definitions could be error-prone.

Let's try to prevent accidents, but retain normal kubectl behavior where possible.